pkg/dbrole/dbrole.go

```diff
@@ -16,55 +16,98 @@
+// that user. If a user has multiple service roles which map to several DB roles,
+// the DB role with the most extensive privileges will be used (see `DbRoles()`
+// for reference to ordered list of DbRoles.
 package dbrole
 
+import "sort"
+
 // DbRole Database roles/users.
 type DbRole string
 
 const (
 	// NO_ROLE DB Roles.
-	NO_ROLE       DbRole = ""
-	TENANT_READER DbRole = "tenant_reader"
-	READER        DbRole = "reader"
-	TENANT_WRITER DbRole = "tenant_writer"
-	WRITER        DbRole = "writer"
-	MAIN          DbRole = "main"
+	NO_ROLE                DbRole = ""
+	TENANT_INSTANCE_READER DbRole = "tenant_instance_reader"
+	TENANT_READER          DbRole = "tenant_reader"
+	INSTANCE_READER        DbRole = "instance_reader"
+	READER                 DbRole = "reader"
+	TENANT_INSTANCE_WRITER DbRole = "tenant_instance_writer"
+	TENANT_WRITER          DbRole = "tenant_writer"
+	INSTANCE_WRITER        DbRole = "instance_writer"
+	WRITER                 DbRole = "writer"
+	MAIN                   DbRole = "main"
 )
 
 // Returns *Ordered* slice of DbRoles.
 // A reader role is always considered to have fewer permissions than a writer role.
 // and a tenant-specific reader/writer role is always considered to have fewer permissions,
 // than a non-tenant specific reader/writer role, respectively.
-// NO_ROLE < TENANT_READER < READER < TENANT_WRITER < WRITER.
 func DbRoles() DbRoleSlice {
 	return DbRoleSlice{
 		NO_ROLE,
+		TENANT_INSTANCE_READER,
 		TENANT_READER,
+		INSTANCE_READER,
 		READER,
+		TENANT_INSTANCE_WRITER,
 		TENANT_WRITER,
+		INSTANCE_WRITER,
 		WRITER,
 		MAIN,
 	}
 }
 
-type DbRoleSlice []DbRole // Needed for sorting records
-func (a DbRoleSlice) Len() int {
-	return len(a)
+func (dbRole DbRole) IsDbRoleTenantScoped() bool {
+	return dbRole == TENANT_READER || dbRole == TENANT_WRITER || dbRole == TENANT_INSTANCE_READER || dbRole == TENANT_INSTANCE_WRITER
 }
 
-func (dbRole DbRole) IsDbRoleTenantScoped() bool {
-	return dbRole == TENANT_READER || dbRole == TENANT_WRITER
+func (dbRole DbRole) IsDbRoleInstanceScoped() bool {
+	return dbRole == INSTANCE_READER || dbRole == INSTANCE_WRITER || dbRole == TENANT_INSTANCE_READER || dbRole == TENANT_INSTANCE_WRITER
+}
+
+// Map roles to instancer based when Instancer is set.
+func (dbRole DbRole) GetRoleWithInstancer() DbRole {
+	switch dbRole {
+	case READER:
+		return INSTANCE_READER
+	case WRITER:
+		return INSTANCE_WRITER
+	case TENANT_READER:
+		return TENANT_INSTANCE_READER
+	case TENANT_WRITER:
+		return TENANT_INSTANCE_WRITER
+	default:
+		return dbRole
+	}
+}
+
+// Map roles to tenancy based roles as tenant column is configured.
+func (dbRole DbRole) GetRoleWithTenancy() DbRole {
+	switch dbRole {
+	case READER:
+		return TENANT_READER
+	case WRITER:
+		return TENANT_WRITER
+	case INSTANCE_READER:
+		return TENANT_INSTANCE_READER
+	case INSTANCE_WRITER:
+		return TENANT_INSTANCE_WRITER
+	default:
+		return dbRole
+	}
```

Comment:
I would have added the `GetRoleWithInstancer` and `GetRoleWithTenancy` methods in a PR where they are actually used, so that it's clear that they are needed

Reply:
Hmm it's a bigger change let me see if we can just pull dbUsers change easily